### PR TITLE
Frontend: Graph Legend Update

### PR DESCRIPTION
**Pull request by ktpa**
[Forked repo discussion](https://github.com/ItJustWorksBetterTM/EiffelVis/pull/21)

Ready for ItJustWorksTM:master ✔️

### CODE REVIEW

**Comment by RuthgerD** [comment]
Haven't had the time to review this just yet, will find some time this week hopefully.
could you maybe post some screenshots of the final product?

**Comment by hry625** [comment]
> Haven't had the time to review this just yet, will find some time this week hopefully. could you maybe post some screenshots of the final product?
截屏2022-11-09 14 35 30
The changes are mostly regards the ColorLegend component. We create a new G6 graph as legend and display the nodes according to the config.json file. 
We also relocate the legend to the top left of the window, and display the legend by default.

File: frontend/src/layout.ts [hunk 1]
```diff
@@ -56,19 +74,44 @@ export class StatefulLayout {
...
+    // if (!this.shapes.has(eventType)) {
+    //   const shape = this.shapeArray[Math.floor(Math.random() * this.shapeArray.length)]
+    //   console.log(shape)
+    //   this.shapes.set(eventType, shape)
+    // }
+    return this.themeMap.get(eventType).Shape
+  }
+  nodeLabel(eventType: string) {
+    // if a new event type are created, we generate a random integer between 0-4 as the index of shape array.
+    // if (!this.shapes.has(eventType)) {
```
Comment (RuthgerD) [inline frontend/src/layout.ts:109]: yeet

File: frontend/src/layout.ts [hunk 2]
```diff
@@ -56,19 +74,44 @@ export class StatefulLayout {
...
+  }
+
+  getNodeStyle() {
+    return this.themeMap
+  }
+  nodeShape(eventType: string) {
+    // if a new event type are created, we generate a random integer between 0-4 as the index of shape array.
+    // if (!this.shapes.has(eventType)) {
+    //   const shape = this.shapeArray[Math.floor(Math.random() * this.shapeArray.length)]
+    //   console.log(shape)
```
Comment (RuthgerD) [inline frontend/src/layout.ts:102]: yeet

File: frontend/src/layout.ts [hunk 3]
```diff
@@ -56,19 +74,44 @@ export class StatefulLayout {
...
-      const hash = [...eventType].reduce(
-        (acc, char) => char.charCodeAt(0) + ((acc << 1) - acc),
-        2
-      )
-      const color = `hsl(${hash % (hue ? hue : 360)},50%,50%)`
-      this.colors.set(eventType, color)
-    }
-    return this.colors.get(eventType)
+  nodeColor(eventType: string) {
+    // if (!this.colors.has(eventType)) {
```
Comment (RuthgerD) [inline frontend/src/layout.ts:78]: yeet

File: frontend/src/layout.ts [hunk 4]
```diff
@@ -16,17 +17,34 @@ export class StatefulLayout {
...
     }
     node.style = {
-      fill: this.nodeColor(node.event_type, graphOptions.hue),
+      fill: this.nodeColor(node.event_type),
       lineWidth: 0.4,
     }
     node.size = 10
+    node.type = this.nodeShape(node.event_type)
+    // comment out labels for the node on the graph
+    // node.label = this.nodeLabel(node.event_type)
```
Comment (RuthgerD) [inline frontend/src/layout.ts:40]: yeet

File: frontend/src/config.txt [hunk 5]
```diff
@@ -0,0 +1,9 @@
+Theme:
```
Comment (RuthgerD) [inline frontend/src/config.txt:1]: What is this file?

File: frontend/src/components/ColorLegend.svelte [hunk 6]
```diff
@@ -1,28 +1,96 @@
...
+
+    let container: HTMLElement;
+    let graph: Graph | null;
+    export const reset = () => {
+        graph?.changeData({});
+        graph?.render();
+        dispatch("nodeselected", null);
+    };
+
+    export const resizeGraph = () => {
```
Comment (RuthgerD) [inline frontend/src/components/ColorLegend.svelte:48]: there seems to be a bit of duplicated code here, could we make the g6 graph into more of a generic component?

File: frontend/src/layout.ts [hunk 7]
```diff
@@ -16,17 +17,34 @@ export class StatefulLayout {
   private curve = 0
   private curveSep = 0
   private colors = new Map<string, string>()
+  private shapes = new Map<string, string>()
+  private shapeArray = ["circle", "rect", "ellipse", "diamond", "star"]
+  private customTheme = config.Theme.ColorBlind
+  private themeMap = new Map(Object.entries(this.customTheme));
+  private defaultColor = this.themeMap.get("EiffelSourceChangeSubmittedEvent").Color
```
Comment (RuthgerD) [inline frontend/src/layout.ts:24]: this seems brittle.

File: frontend/src/layout.ts [hunk 8]
```diff
@@ -56,19 +74,44 @@ export class StatefulLayout {
...
-    return this.colors.get(eventType)
+  nodeColor(eventType: string) {
+    // if (!this.colors.has(eventType)) {
+    //   const hash = [...eventType].reduce(
+    //     (acc, char) => char.charCodeAt(0) + ((acc << 1) - acc),
+    //     2
+    //   )
+    //   const color = `hsl(${hash % (hue ? hue : 360)},50%,50%)`
+    //   this.colors.set(eventType, color)
+    return this.themeMap.get(eventType).Color
```
Comment (RuthgerD) [inline frontend/src/layout.ts:85]: we need some error handling here in the case event type is not in the theme

**Review by RuthgerD — CHANGES_REQUESTED** [review]
Not sure how I feel about this one, but clean up the code a bit first.

File: frontend/src/layout.ts (hunk 8 above)
Comment (alrapal) [reply frontend/src/layout.ts:85]: The person who has been working on this is currently sick but we will forward the feedback so it is worked on when they are feeling better =)

**Comment by hry625** [comment]
I have clean up the code in the layout.ts file. It seems some late changes cause some conflicts to resolve.

**Comment by hry625** [comment]
We resolve the conflicts and we put the legend and the button of the legend back to the original place. We will change the location of the panel in other branch. The result is shown as follow.

Screenshot 2022-11-22 at 16 46 01

The only thing we are not sure is that the undefined type is called "Unknown" in the legend right now, but it is too long to display. We are suggest to change it to "Unk." for short. What's your opinion about that?

**Comment by RuthgerD** [comment]
> We are suggest to change it to "Unk." for short. What's your opinion about that?

`Unk.` is not really a common term people use I think, try making it a quesiton mark `?`, or even `???`; that will visually tell people what happened.

**Comment by RuthgerD** [comment]
The legend looks good as shown in the pic, if we can change the orange header to a more muted color (with white text) then Im happy to include this change.

**Comment by alrapal** [comment]
> > We are suggest to change it to "Unk." for short. What's your opinion about that?
> 
> `Unk.` is not really a common term people use I think, try making it a quesiton mark `?`, or even `???`; that will visually tell people what happened.

Good idea with the `?` ! Maybe one is enough ?

**Comment by hry625** [comment]
> The legend looks good as shown in the pic, if we can change the orange header to a more muted color (with white text) then Im happy to include this change.

I like the ??? solution.

**Comment by hry625** [comment]
截屏2022-11-22 19 12 27

Dose this color works for you?

**Comment by RuthgerD** [comment]
Not bad actually! however I think if the overarching plan is to put the legend in the left card stack it shoudnt have an eye catching header like that, it needs to be a card like the current filter panel. If the legend stays like that im happy with it.

**Comment by alrapal** [comment]
The plan is still to have the side bar, yes. 
I think the purple is a bit strange to have since there is no other element that is purple. Maybe, if orange is too flashy, we could go for the light gray that is on the buttons when hovered? 
We will most likely need something consistent between panels anyway

**Comment by RuthgerD** [comment]
Yeah just stay consistent with gray

**Comment by hry625** [comment]
截屏2022-11-23 11 58 27
Now the background color of legend title is set to gray.

**Comment by RuthgerD** [comment]
Meant dark gray to match the rest, not light gray :see_no_evil:

**Comment by hry625** [comment]
> Meant dark gray to match the rest, not light gray 🙈
The previous one is ```bg-gray-500```
And ```bg-gray-300``` is even lighter.
截屏2022-11-23 12 40 58
```bg-gray-700``` is more blue-ish to me.
截屏2022-11-23 12 43 00

**Comment by RuthgerD** [comment]
Look here: https://daisyui.com/docs/colors/

**Comment by kam56** [comment]
> > Meant dark gray to match the rest, not light gray 🙈
> 
> The previous one is `bg-gray-500` And `bg-gray-300` is even lighter. 截屏2022-11-23 12 40 58
> 
> `bg-gray-700` is more blue-ish to me. 截屏2022-11-23 12 43 00

Hey Renyaun, I have used `bg-base-200` for the filter panel header. And, that matches perfectly the rest of the elements. I think in the version that you are working with, `bg-base-100` is the best color.

**Comment by kam56** [comment]
We will update the colors slightly in this [issue](https://github.com/ItJustWorksBetterTM/EiffelVis/pull/40 ). So, I suggest that we use `bg-base-100` for now.

**Comment by hry625** [comment]
截屏2022-11-24 12 12 59
I made the title color consistent with filter panel.

**Comment by RuthgerD** [comment]
That works for me

File: frontend/package.json [hunk 9]
```diff
@@ -34,4 +34,4 @@
     "style-loader": "^3.3.1",
     "tailwindcss": "^3.0.12"
   }
-}
+}
```
Comment (RuthgerD) [inline frontend/package.json:37]: unneeded change here

File: frontend/src/components/ColorLegend.svelte [hunk 10]
```diff
@@ -1,28 +1,114 @@
...
+        labelCfg: {
+            style: {
+                fill: defaultNode.color,
+            },
+            position: "bottom",
+            offset: 10,
+        },
+    };
+    array.push(dNode);
+    data = { nodes: array };
```
Comment (RuthgerD) [inline frontend/src/components/ColorLegend.svelte:56]: Please make this a function instead, it is not easy to understand for as to what this is doing.

File: frontend/src/components/ColorLegend.svelte [hunk 11]
```diff
@@ -1,28 +1,114 @@
...
+                style: {
+                    fill: event[1].Color,
+                },
+                position: "bottom",
+                offset: 10,
+            },
+        };
+        array.push(node);
+    });
+    let dNode = {
```
Comment (RuthgerD) [inline frontend/src/components/ColorLegend.svelte:38]: dNode -> default_node, to be descriptive and snake_case

File: frontend/package.json (hunk 9 above)
Comment (hry625) [reply frontend/package.json:37]: I think it's the auto indentation cause the problem.

**Comment by RuthgerD** [comment]
Ready for a last review?

**Comment by hry625** [comment]
> Ready for a last review?

Yes, let's go

**Review by RuthgerD — APPROVED** [review]

**Comment by RuthgerD** [comment]
Thanks!

**Pull request merged**